Add Pardiso solver for SparseMatrix

adam-sim-dev opened the pull request:
This PR adds MKL Pardiso solver for SparseMatrix. Please

1) Format the code style.
2) Check if there are other files I need to update. (e.g., in the doc).
3) Review and test carefully.
4) I suggest updating `MKL_LIBRARY_DIR` to "lib/intel64".

Best,
Adam<!--GHEX{"author":"adam-sim-dev","assignment":"2023-06-11T22:26:33","editor":"tzanio","id":3615,"reviewers":["jandrej","cjvogl","psocratis"],"merge":"2023-07-10T18:36:23","approval":"2023-07-10T18:35:40"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3615](https://github.com/mfem/mfem/pull/3615) | @adam-sim-dev | @tzanio | @jandrej + @cjvogl + @psocratis | 6/11/23 | 7/10/23 | 7/10/23 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

v-dobrev commented:
Hi @adam-sim-dev, please add the `ready-for-review` label when this PR is ready for review.

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

cjvogl commented:
I should be able to do a code review in the next few days.  @adam-sim-dev, can you suggest a test to run (or even better yet, add a unit test to the MR)?

cjvogl commented on tests/unit/linalg/test_direct_solvers.cpp:
```diff
@@ -138,19 +141,40 @@ TEST_CASE("Serial Direct Solvers", "[CUDA]")
-      REQUIRE(Y.Norml2() < 1.e-12);
+         a.RecoverFEMSolution(X, b, x);
+         VectorFunctionCoefficient grad(dim, gradexact);
+         double error = x.ComputeH1Error(&uex, &grad);
```
This variable is goes out of scope on the following line, so there should be a compiler error on 154/178.  When I compile with cmake, I don't see the error because `MFEM_USE_MKL_PARDISO` is not defined, so `DIRECT_SOLVE_SERIAL` is also not defined.  When I build with SuiteSparse (so that `DIRECT_SOLVE_SERIAL` is defined), I get the expected compilation error.

@adam-sim-dev, do you use the makefile or cmake build system?

cjvogl replied:
To fix the cmake build system, you can apply the following patch
```git
diff --git a/CMakeLists.txt b/CMakeLists.txt
index e9e6ae0..7e30e5c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -473,6 +473,10 @@ if (MFEM_USE_MKL_CPARDISO)
   endif()
 endif()
 
+if (MFEM_USE_MKL_PARDISO)
+  find_package(MKL_PARDISO REQUIRED MKL_SEQUENTIAL MKL_LP64 MKL_MPI_WRAPPER)
+endif()
+
 # PARELAG
 if (MFEM_USE_PARELAG)
    find_package(PARELAG REQUIRED)
@@ -522,8 +526,8 @@ find_package(Threads REQUIRED)
 set(MFEM_TPLS OPENMP HYPRE LAPACK BLAS SuperLUDist STRUMPACK METIS SuiteSparse
     SUNDIALS PETSC SLEPC MUMPS AXOM FMS CONDUIT Ginkgo GNUTLS GSLIB
     NETCDF MPFR PUMI HIOP POSIXCLOCKS MFEMBacktrace ZLIB OCCA CEED RAJA UMPIRE
-    ADIOS2 CUBLAS CUSPARSE MKL_CPARDISO AMGX CALIPER CODIPACK BENCHMARK PARELAG
-    MPI_CXX HIP HIPSPARSE MOONOLITH BLITZ ALGOIM ENZYME)
+    ADIOS2 CUBLAS CUSPARSE MKL_CPARDISO MKL_PARDISO AMGX CALIPER CODIPACK
+    BENCHMARK PARELAG MPI_CXX HIP HIPSPARSE MOONOLITH BLITZ ALGOIM ENZYME)
 
 # Add all *_FOUND libraries in the variable TPL_LIBRARIES.
 set(TPL_LIBRARIES "")
```

To fix the unit test, you can apply the following patch
```bash
diff --git a/tests/unit/linalg/test_direct_solvers.cpp b/tests/unit/linalg/test_direct_solvers.cpp
index 2c40952..48a1ac3 100644
--- a/tests/unit/linalg/test_direct_solvers.cpp
+++ b/tests/unit/linalg/test_direct_solvers.cpp
@@ -156,6 +156,7 @@ TEST_CASE("Serial Direct Solvers", "[CUDA]")
          a.RecoverFEMSolution(X, b, x);
          VectorFunctionCoefficient grad(dim, gradexact);
          double error = x.ComputeH1Error(&uex, &grad);
+         REQUIRE(error < 1.e-12);
       }
 #endif
 #ifdef MFEM_USE_MKL_PARDISO
@@ -172,10 +173,9 @@ TEST_CASE("Serial Direct Solvers", "[CUDA]")
          a.RecoverFEMSolution(X, b, x);
          VectorFunctionCoefficient grad(dim, gradexact);
          double error = x.ComputeH1Error(&uex, &grad);
+         REQUIRE(error < 1.e-12);
       }
 #endif
-
-      REQUIRE(error < 1.e-12);
    }
 }
```

cjvogl reviewed (changes requested):
Look pretty good overall!  Just needs a few things addressed:

- [x] Add Pardiso dependency to `CMakeLists.txt` (see patch provided in inline comment)
- [x] Fix unit test (see patch provided in inline comment)
- [x] Confirm CMake issue does not also exist in makefile build system

adam-sim-dev commented:
Thanks @cjvogl for the review! I have addressed all your comments. While I think MKL_MPI_WRAPPER is not needed for MKL_PARDISO. I have tested 1) on Windows with CMake, Intel oneAPI and Visual Studio, 2) on Ubuntu with CMake, Makefile, and Intel oneAPI.

tzanio commented:
@sshiraiwa, can you take a look?

cjvogl commented:
> I think MKL_MPI_WRAPPER is not needed for MKL_PARDISO

@adam-sim-dev, that's a good point!  I admittedly just copy-pasted the line from CPardiso

cjvogl reviewed (approved):
Reran everything with a cmake build: looks good to me

psocratis commented on linalg/pardiso.cpp:
```diff
@@ -0,0 +1,146 @@
+
+   nnz = mat->NumNonZeroElems();
+
+   if (mat->IsSymmetric() == 0.0)
```
This check might be unnecessarily expensive since it will parse all the entries of the matrix to check for symmetry. 
Maybe allowing the user to set the symmetry would suffice.

jandrej replied:
I agree to remove this

cjvogl replied:
Same: leave it to the use to specify if symmetric, good catch @psocratis

psocratis reviewed (approved):
This looks good. Unit and Valgrind tests are successful. Very nice addition!

tzanio commented:
Merged in `next` for testing...

tzanio commented:
Please mention in `CHANGELOG`

tzanio commented on CHANGELOG:
```diff
@@ -60,6 +60,9 @@ Linear and nonlinear solvers
   sides, block low-rank compression, builds using 64-bit integers, and other
   improvements.
 
+- Added an interface to the MKL Pardiso solver, a shared-memory multiprocessing
+  parallel direct sparse solver developed by Intel.
```
Thanks for adding this.

The comment for `MKL CPardiso` from mfem-4.2 was:

> Added an interface to the MKL CPardiso solver, an MPI-parallel sparse direct
> solver developed by Intel. See Example 11p for an illustration of its usage.

The above sounds very similar to it, can we clarify what the difference between the two solvers is in the above comment?

cc: @jandrej, @psocratis

jandrej replied:
How about

> Added an interface to the MKL Pardiso solver, a shared-memory multiprocessing
> parallel direct sparse solver developed by Intel. This interface provides a serial version of Pardiso for use with `SparseMatrix`.

tzanio replied:
How does `multiprocessing parallel` connect  with `SparseMatrix`, which is serial?

tzanio replied:
```suggestion
- Added an interface to the MKL Pardiso sparse direct solver developed by Intel.
  This interface provides a serial (OpenMP shared memory) version of Pardiso for
  use with SparseMatrix. This complements the existing parallel (MPI distributed
  memory) version already available through the CPardiso MFEM integration.
```

@adam-sim-dev and @jandrej -- are you OK with this rewording?

adam-sim-dev replied:
Looks good to me.

jandrej replied:
👍

tzanio commented:
Re-merged in `next` for testing...

rfalko commented:
Hello, I have a question about this PR. I tested this solver and got an error because the solver always reports 0 height and width. Maybe a

height = op.Height();
width = op.Width();

is missing in the SetOperator call?

psocratis commented:
> Hello, I have a question about this PR. I tested this solver and got an error because the solver always reports 0 height and width. Maybe a
> 
> height = op.Height(); width = op.Width();
> 
> is missing in the SetOperator call?

Good catch! @adam-sim-dev ?

tzanio commented:
Re-merged in `next` for testing...

adam-sim-dev commented:
Does mfem-4.5\config\cmake\config.hpp.in need an update too?

jandrej commented:
I didn't check for make compatibility. A simple
```
// Enable interface to the MKL Pardiso library.
#cmakedefine MFEM_USE_MKL_PARDISO
```
should work with the existing setup

tzanio commented:
@adam-sim-dev, can you open a quick bugfix PR with the above line?

adam-sim-dev commented:
And config/cmake/MFEMConfig.cmake.in? MFEM_USE_MKL_CPARDISO is missing too.

tzanio commented:
> And config/cmake/MFEMConfig.cmake.in? MFEM_USE_MKL_CPARDISO is missing too.

Yes
